src/twodtree.h

```diff
@@ -0,0 +1,108 @@
+// Copyright 2025 The Manifold Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+
+#include "./utils.h"
+#include "manifold/common.h"
+#include "manifold/optional_assert.h"
+#include "manifold/vec_view.h"
+
+namespace manifold {
+
+template <typename P, bool sortX = true>
+void BuildTwoDTreeImpl(VecView<P> points) {
+  std::sort(points.begin(), points.end(), [](const P& a, const P& b) {
+    return sortX ? a.pt.x < b.pt.x : a.pt.y < b.pt.y;
+  });
+  if (points.size() < 2) return;
+  BuildTwoDTreeImpl<P, !sortX>(points.view(0, points.size() / 2));
+  BuildTwoDTreeImpl<P, !sortX>(points.view(points.size() / 2 + 1));
+}
+
+// Not really a proper KD-tree, but a kd tree with k = 2 and alternating x/y
+// partition.
+// Recursive sorting is not the most efficient, but simple and guaranteed to
+// result in a balanced tree.
+//
+// Make it multiway...
+template <typename P>
+void BuildTwoDTree(VecView<P> points) {
+  ZoneScoped;
+  // don't even bother...
+  if (points.size() <= 8) return;
+  BuildTwoDTreeImpl<P, true>(points);
+}
+
+template <typename P, typename F>
+void QueryTwoDTree(VecView<P> points, Rect r, F f) {
```

Comment:
So, if this is faster than the collider for 2D, do you think a version of it will also be faster for 3D? It would be nice to have something universal. Or failing that, to at least have them use the same kind of API so it's easier to compare them? Is that reasonable?

Reply:
I need to think a bit about this.